### PR TITLE
Feat add controller

### PR DESCRIPTION
**Pull request by antonio-farina**
Hi Dam!

I created some tests for the ElementController.

### CODE REVIEW

File: src/main/java/com/github/metabs/model/Element.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import com.github.metabs.model.vo.converter.DescriptionConverter;
+import com.github.metabs.model.vo.converter.NameConverter;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.UUID;
+import org.springframework.data.neo4j.core.convert.ConvertWith;
+import org.springframework.data.neo4j.core.schema.Id;
+import org.springframework.data.neo4j.core.schema.Relationship;
+
+@org.springframework.data.neo4j.core.schema.Node("Element")
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/model/Element.java:14]: I think we can add this line to the import block :)

File: src/main/java/com/github/metabs/model/dto/SaveElementDto.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package com.github.metabs.model.dto;
+
+import com.github.metabs.model.vo.Description;
+import com.github.metabs.model.vo.Name;
+import java.net.URL;
+
+
+public class SaveElementDto {
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/model/dto/SaveElementDto.java:8]: Why is this inside the model package? I think we should move it inside the service package

File: src/main/java/com/github/metabs/model/dto/SaveElementRequest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package com.github.metabs.model.dto;
+
+public class SaveElementRequest {
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/model/dto/SaveElementRequest.java:3]: Why is this inside the model package? I think we should move it inside the controller/request package (or similar)

File: src/main/java/com/github/metabs/model/dto/validator/RequestElementValidator.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import com.github.metabs.model.exception.NameException;
+import com.github.metabs.model.vo.Description;
+import com.github.metabs.model.vo.Name;
+import java.net.MalformedURLException;
+import java.net.URL;
+import org.springframework.validation.Errors;
+import org.springframework.validation.Validator;
+
+//@Component
+public class RequestElementValidator implements Validator {
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/model/dto/validator/RequestElementValidator.java:15]: Why is this inside the model package? I think we should move it inside the controller/validator package

File: src/main/java/com/github/metabs/model/vo/converter/LoggerDecoratorConverter.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package com.github.metabs.model.vo.converter;
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/model/vo/converter/LoggerDecoratorConverter.java:1]: For the moment we can delete this if you want :)

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Optional;
+import java.util.UUID;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ElementService {
+
+  @Autowired
+  ElementRepository elementRepository;
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:19]: Add visibility

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+  @Autowired
+  CollectionRepository collectionRepository;
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:22]: Add visibility

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  public ElementService() {
+  }
+
+  public Optional<Element> getElementById(UUID id) {
+    return elementRepository.findById(id);
+  }
+
+  public Element saveElement(SaveElementDto saveSaveElementDtoRequestDto) {
+
+    com.github.metabs.model.Element element;
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:33]: We can reuse the import block for this :)

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  CollectionRepository collectionRepository;
+
+  public ElementService() {
+  }
+
+  public Optional<Element> getElementById(UUID id) {
+    return elementRepository.findById(id);
+  }
+
+  public Element saveElement(SaveElementDto saveSaveElementDtoRequestDto) {
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:31]:
```suggestion
  public Element saveElement(SaveElementDto dto) {
```

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    com.github.metabs.model.Element element;
+    if (saveSaveElementDtoRequestDto.getLink() == null) {
+      UUID id = UUID.randomUUID();
+      element = Collection.createCollection(
+          id,
+          saveSaveElementDtoRequestDto.getName(),
+          saveSaveElementDtoRequestDto.getDescription()
+      );
+    } else {
+      UUID id2 = UUID.randomUUID();
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:42]: Can't we create the `id` variable only once before the first if statement and reuse it?

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+          id2,
+          saveSaveElementDtoRequestDto.getName(),
+          saveSaveElementDtoRequestDto.getLink(),
+          saveSaveElementDtoRequestDto.getDescription()
+      );
+    }
+    return elementRepository.save(element);
+  }
+
+  public com.github.metabs.model.Element saveElementWithParent(
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:53]: We can reuse the import block for this :)

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+      SaveElementDto saveSaveElementDtoRequestDto,
+      UUID parentCollectionId
+  ) throws ParentNotFoundException {
+    Optional<Collection> parentCollection = collectionRepository.findById(parentCollectionId);
+    if (!parentCollection.isPresent()) {
+      throw ParentNotFoundException.parentNotFound();
+    }
+    com.github.metabs.model.Element element;
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:61]: We can reuse the import block for this :)

File: src/main/java/com/github/metabs/service/ElementService.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+      element = Collection.createCollectionWithParent(
+          id,
+          parentCollection.get(),
+          saveSaveElementDtoRequestDto.getName(),
+          saveSaveElementDtoRequestDto.getDescription()
+      );
+    } else {
+      UUID id2 = UUID.randomUUID();
+      element = Tab.createTabWithParent(
+          id2,
```
Comment (damianopetrungaro) [inline src/main/java/com/github/metabs/service/ElementService.java:73]: Can't we create the id variable only once before the first if statement and reuse it?

File: src/test/java/com/github/metabs/controller/ElementControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+  private RequestElementValidator validator;
+
+  @Mock
+  private Errors errors;
+
+  @InjectMocks
+  private ElementController elementController;
+
+  @Before
+  public void setup() {
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/controller/ElementControllerTest.java:61]: We should consider moving this inside a configuration for the tests, but maybe not in this PR

File: src/test/java/com/github/metabs/controller/ElementControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+
+    Mockito.when(
+        elementService.getElementById(tab.getId())
+    ).thenReturn(Optional.of(tab));
+
+    RequestBuilder request = MockMvcRequestBuilders.
+        get("/elements/" + tab.getId().toString()).
+        accept(MediaType.APPLICATION_JSON);
+
+    MvcResult result = mockMvc.perform(request).andReturn();
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/controller/ElementControllerTest.java:93]: Let's use the assertions provided by mockMvc

File: src/test/java/com/github/metabs/controller/ElementControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+        )
+        .andExpect(jsonPath("$.name").value(collection.getName().getValue()))
+        .andExpect(jsonPath("$.description").value(collection.getDescription().getValue()))
+        .andExpect(jsonPath("$.created").value(collection.getCreated().toString()))
+        .andExpect(jsonPath("$.updated").doesNotExist())
+        .andExpect(jsonPath("$.trashed").doesNotExist())
+        .andExpect(status().isCreated());
+  }
+/**
+ @Test public void shouldReturnNonProcessableEntity() throws Exception {
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/controller/ElementControllerTest.java]: Add a test for a 422

File: src/test/java/com/github/metabs/controller/serialzer/CollectionJsonSerializerTest.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+@JsonTest
+public class CollectionJsonSerializerTest {
+
+  @Autowired
+  private JacksonTester<Collection> jacksonTester;
+
+  @Test
+  public void serializeFromJavaToJson() throws Exception {
+
+    UUID idParent = UUID.randomUUID();
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/controller/serialzer/CollectionJsonSerializerTest.java:26]: User the object mother to create collection

File: src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
+package com.github.metabs.model.dto;
+
+import com.github.metabs.model.ObjectMother;
+import com.github.metabs.model.vo.Description;
+import com.github.metabs.model.vo.Name;
+import java.net.URL;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class SaveSaveElementDtoRequestDtoTest {
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java:10]: Change all the names after you moved again the src class

File: src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestTest.java [hunk 19]
```diff
@@ -0,0 +1,45 @@
+package com.github.metabs.model.dto;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class SaveSaveElementDtoRequestTest {
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestTest.java:6]: Change all the names after you moved again the src class

File: src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+
+  SaveElementDto saveElementDto = new SaveElementDto();
+
+  @Test
+  public void shouldRename() throws Exception {
+    Name name = ObjectMother.generateRandomName();
+    saveElementDto.rename(name);
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java:17]: change to get/set for those methods, the rename are amazing for domain objects, but for DTOs I suggest you to keep them more simple

File: src/test/java/com/github/metabs/service/ElementServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,6 @@
+package com.github.metabs.service;
+
+public class ElementServiceTest {
+
+  //TODO
```
Comment (damianopetrungaro) [inline src/test/java/com/github/metabs/service/ElementServiceTest.java:5]: To add after this PR get merged

**Review by damianopetrungaro — CHANGES_REQUESTED** [review]
Hey! 
Well done, just commented some things I've seen :) 

Let me know what you think!

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("elements")
+public class ElementController {
+
+  @Autowired
+  ElementService elementService;
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java]: Done

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+
+  @Resource(name = "requestElementValidator")
+  RequestElementValidator validator;
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java]: Done

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+
+  @GetMapping("/{id}")
+  public ResponseEntity<Element> getElementById(@PathVariable UUID id) {
+    try {
+      Optional<Element> element = elementService.getElementById(id);
+      if (!element.isPresent()) {
+        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+      }
+      Element trueElement = element.get();
+      return new ResponseEntity<>(trueElement, HttpStatus.OK);
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java:42]: Done

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+    }
+  }
+
+  @PostMapping("/")
+  @ResponseBody
+  public ResponseEntity<Object> saveElement(@RequestBody SaveElementRequest saveElementRequest,
+                                            BindingResult result) {
+
+    validator.validate(saveElementRequest, result);
+    SaveElementDto saveElementDto = validator.getSaveElementDto();
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java]: Done

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+
+  @PostMapping("/{id}")
+  @ResponseBody
+  public ResponseEntity<Object> saveElementWithParent(
+      @RequestBody SaveElementRequest saveElementRequest,
+      @PathVariable("id") UUID parentCollectionId,
+      BindingResult result
+  ) {
+    validator.validate(saveElementRequest, result);
+    SaveElementDto saveElementDto = validator.getSaveElementDto();
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java]: Done

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+      if (!element.isPresent()) {
+        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+      }
+      elementService.trashElementById(id);
+      return new ResponseEntity<>(HttpStatus.OK);
+    } catch (Exception ex) {
+      return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java:112]: Done

File: src/main/java/com/github/metabs/controller/serializer/CollectionJsonSerializer.java [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+  @Override
+  public void serialize(
+      Collection value,
+      JsonGenerator gen,
+      SerializerProvider serializers
+  ) throws IOException {
+
+    gen.writeStartObject();
+    gen.writeStringField(
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/serializer/CollectionJsonSerializer.java:21]: Done

File: src/main/java/com/github/metabs/controller/serializer/TabJsonSerializer.java [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.github.metabs.model.Tab;
+import java.io.IOException;
+import org.springframework.boot.jackson.JsonComponent;
+
+@JsonComponent
+public class TabJsonSerializer extends JsonSerializer<Tab> {
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/serializer/TabJsonSerializer.java:11]: Done

File: src/main/java/com/github/metabs/model/Element.java (hunk 1 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/model/Element.java:14]: Done

File: src/main/java/com/github/metabs/model/dto/SaveElementDto.java (hunk 2 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/model/dto/SaveElementDto.java:8]: Done

File: src/main/java/com/github/metabs/model/dto/SaveElementRequest.java (hunk 3 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/model/dto/SaveElementRequest.java:3]: Done

File: src/main/java/com/github/metabs/model/dto/validator/RequestElementValidator.java (hunk 4 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/model/dto/validator/RequestElementValidator.java:15]: Done

File: src/main/java/com/github/metabs/model/vo/converter/LoggerDecoratorConverter.java (hunk 5 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/model/vo/converter/LoggerDecoratorConverter.java:1]: Done

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 6 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java:19]: Done

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 7 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java:22]: Done

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 8 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java:33]: Yes we can!

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 9 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java]: Ok

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 10 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java:42]: Yes we can!

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 11 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java:53]: Yes!

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 12 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java:61]: Right!

File: src/main/java/com/github/metabs/service/ElementService.java (hunk 13 above)
Comment (antonio-farina) [reply src/main/java/com/github/metabs/service/ElementService.java:73]: Right!

File: src/test/java/com/github/metabs/controller/ElementControllerTest.java (hunk 14 above)
Comment (antonio-farina) [reply src/test/java/com/github/metabs/controller/ElementControllerTest.java:53]: ok we will talk about it

File: src/test/java/com/github/metabs/controller/serialzer/CollectionJsonSerializerTest.java (hunk 17 above)
Comment (antonio-farina) [reply src/test/java/com/github/metabs/controller/serialzer/CollectionJsonSerializerTest.java:38]: Done

File: src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java (hunk 18 above)
Comment (antonio-farina) [reply src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java:10]: Done

File: src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestTest.java (hunk 19 above)
Comment (antonio-farina) [reply src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestTest.java:6]: Done

File: src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java (hunk 20 above)
Comment (antonio-farina) [reply src/test/java/com/github/metabs/model/dto/SaveSaveElementDtoRequestDtoTest.java:17]: Done

File: src/test/java/com/github/metabs/service/ElementServiceTest.java (hunk 21 above)
Comment (antonio-farina) [reply src/test/java/com/github/metabs/service/ElementServiceTest.java:5]: ok we will do it later

**Review by damianopetrungaro — CHANGES_REQUESTED** [review]
You forgot a couple of those :D

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+    }
+  }
+
+  @PostMapping("/")
+  @ResponseBody
+  public ResponseEntity<Object> saveElement(
+      @RequestBody SaveElementRequest elementRequest,
+      BindingResult result
+  ) {
+    if (result.hasErrors()) {
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java:52]: ok

File: src/main/java/com/github/metabs/controller/ElementController.java [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+    } catch (Exception ex) {
+      return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  @DeleteMapping("/{id}")
+  public ResponseEntity<Object> trashElementById(
+      @PathVariable("id") UUID id) {
+    try {
+      Optional<Element> element = elementService.getElementById(id);
```
[the remark this answers is not in the document]
Comment (antonio-farina) [reply src/main/java/com/github/metabs/controller/ElementController.java:96]: Ok

**Review by damianopetrungaro — APPROVED** [review]

**Comment by antonio-farina** [comment]
feat:add element RESTAPI

**Pull request merged**